File: Composition/Java/Composition/src/main/java/algorithm/composition/PointsAggregator.java
package algorithm.composition;

import java.util.ArrayList;
import java.util.Collection;

import algorithm.Measurement;

public class PointsAggregator {

	protected final Collection<Measurement> measurements;
	private final IMeasurementFilter filter;
	private final IAggregationStrategy aggregator;

	public PointsAggregator(Collection<Measurement> measurements, IMeasurementFilter filter,
			IAggregationStrategy aggregator) {
		this.measurements = measurements;
		this.filter = filter;
		this.aggregator = aggregator;
	}

	public Measurement aggregate() {
		Collection<Measurement> local = new ArrayList<Measurement>(measurements);
		local = filter.filter(local);
		return aggregator.aggregate(local);
	}

}

File: Composition/CS/Composition/Algorithm/Composition/PointsAggregator.cs
using System.Collections.Generic;

namespace Algorithm.Composition
{
    public class PointsAggregator
    {
        public PointsAggregator(
            IEnumerable<Measurement> measurements, 
            IMeasurementFilter filter, 
            IAggregationStrategy aggregator)
        {
            _measurements = measurements;
            _filter = filter;
            _aggregator = aggregator;
        }

        public virtual Measurement Aggregate()
        {
            var measurements = _measurements;
            measurements = _filter.Filter(measurements);            
            return _aggregator.Aggregate(measurements);
        }

        private readonly IEnumerable<Measurement> _measurements;
        private readonly IMeasurementFilter _filter;
        private readonly IAggregationStrategy _aggregator;
    }        
}